lib/nostrum/struct/emoji.ex

```diff
@@ -3,30 +3,111 @@ defmodule Nostrum.Struct.Emoji do
   Struct representing a Discord emoji.
   """
 
+  alias Nostrum.Struct.Snowflake
+  alias Nostrum.Struct.User
+  alias Nostrum.Util
+
+  defstruct [
+    :id,
+    :name,
+    :user,
+    :require_colons,
+    :managed,
+    :animated,
+    :roles
+  ]
+
+  @typedoc ~S"""
+  Emoji string to be used with the Discord API.
+
+  Some API endpoints take an `emoji`. If it is a custom emoji, it must be 
+  structured as `"id:name"`. If it is an unicode emoji, it can be structured 
+  as any of the following: 
+
+    * `"name"`
+    * A base 16 unicode emoji string.
+    * A URI encoded string.
+
+  We suggest library users to use the `to_api_name/1` function to get a 
+  `Nostrum.Struct.Emoji`'s emoji api name.
+
+  ## Examples
+
+      # Custom Emojis
+      "nostrum:431890438091489"
+
+      # Unicode Emojis
+      "≡ƒæì"
+      "\xF0\x9F\x98\x81"
+      URI.encode("\u2b50")
+
+      # All Emojis
+      emoji = %Nostrum.Struct.Emoji{id: 43819043108, name: "foxbot"}
+      Nostrum.Struct.Emoji.to_api_name(emoji)
+      "foxbot:43819043108"
+
+  """
+  @type emoji_api_name :: String.t
+
   @typedoc "Id of the emoji"
-  @type id :: integer
+  @type id :: Snowflake.t | nil
 
   @typedoc "Name of the emoji"
   @type name :: String.t
 
-  @typedoc """
-  Custom emoji string to be used with API
+  @typedoc "Roles this emoji is whitelisted to"
+  @type roles :: [Snowflake.t] | nil
 
-  Some API endpoints take an `emoji`.
-  The emoji should be structured as `"id:name"`.
-  """
-  @type custom_emoji :: String.t
+  @typedoc "User that created this emoji"
+  @type user :: User.t | nil
+
+  @typedoc "Whether this emoji must be wrapped in colons "
+  @type require_colons :: boolean | nil
+
+  @typedoc "Whether this emoji is managed"
+  @type managed :: boolean | nil
+
+  @typedoc "Whether this emoji is animated"
+  @type animated :: boolean | nil
 
   @type t :: %__MODULE__{
     id: id,
-    name: name
+    name: name,
+    roles: roles,
+    user: user,
+    require_colons: require_colons,
+    managed: managed,
+    animated: animated
   }
 
-  @derive [Poison.Encoder]
-  defstruct [
-    :id,
-    :name
-  ]
+  @doc """
+  Formats an emoji struct into its respective markdown.
+
+  ## Examples
+
+      emoji = %Nostrum.Struct.Emoji{id: 43819043108, name: "foxbot"}
+      Nostrum.Struct.Emoji.to_markdown(emoji)
+      "<:foxbot:43819043108>"
+  """
+  @spec to_markdown(t) :: String.t
+  def to_markdown(emoji)
```

Comment:
For these two functions, do we want to include a function definition that bypasses having to supply a struct?

I'm thinking just a 2 arity function of the same name that takes an `id` and a `name`. Or maybe a 3 arity with default animated set to `false`?

Maybe the use case is kind of niche? Let me know what you think.

Reply:
Seems niche IMO. Even if I were to add it, I'd prefer to create a new module for that. Functions in `Nostrum.Struct.Emoji` should explicitly deal with the Emoji struct and nothing else IMO. However, we could add this for backwards compatibility. I'll think about it.

With that being said, I'll probably allow maps (in addition to the Emoji struct) to work with this function. Just now, I realized that parts of the library aren't properly decoded to their specific structs. Functions like these will break if I don't allow maps to work.

Reply:
If the payload isn't being decoded properly (which is the case a lot of the time, probably), then that is an issue and I don't think we should support that. I partly agree with your idea that we should keep it limited to functions dealing with Emoji structs. If we're going down this route we should stick to it and not allow maps.

I could see users already knowing the id and name and wanting to bypass having to create a struct, but in that case I guess they can just form their own string?

Reply:
I'll revert the map changes then.

For the idea you proposed with having an `id` and `name`, I have a solution, but it might be hard to implement right now. It also suggests having a rather strong reliance on using structs in the project:

In my current project, I currently bind emoji markdown to function names for ease of access:

```Elixir
defmodule Emojis do
  def unranked, do: "<:unranked:425713028852350976>"
  def silver, do: "<:silver:425713020862332930>"
end

# Example

"This text will be sent with this emoji: #{Emojis.silver()}"
"This text will be sent with this emoji: <:silver:425713020862332930>"
```

I was thinking in the future, we could have something like an `EmojiCache` to select emojis. Assume `EmojiCache.get/1` accepts a keyword list and returns a `Nostrum.Struct.Emoji` or `nil`:

```Elixir
defmodule Emojis do
  alias Nostrum.Cache.EmojiCache

  def unranked, do: EmojiCache.get(name: "unranked", id: 425713028852350976)
  def silver, do: EmojiCache.get(name: "silver", id: 425713020862332930)
end

# Example

Emojis.silver()
%Nostrum.Struct.Emoji{
  id: 425713020862332930,
  name: "silver",
  roles: []
}

# Assuming `Nostrum.Struct.Emoji` will implement the `String.Chars` protocol, 
# this will be possible:

"This text will be sent with this emoji: #{Emojis.silver()}"
"This text will be sent with this emoji: <:silver:425713020862332930>"

# As well as pulling information wherever necessary:

Emojis.silver().name
"silver"

Emojis.silver().roles
[]
```

Having an `EmojiCache` would also open users up to error handling in case an emoji is not there:

```Elixir
alias Nostrum.Cache.EmojiCache

case EmojiCache.get(id: 4839104319) do
  nil -> "\:silver\:"  # In case the emoji got deleted
  emoji -> emoji
end
```

While this next positive might be beyond the scope of nostrum, we could also open up this `EmojiCache` to fetch an emoji via markdown:

```Elixir
"This text will be sent with this emoji: <:silver:425713020862332930>"
|> String.split(" ")
|> Enum.each(fn token -> 
  case token do
    "<:" <> _rest = emoji_token -> 
      EmojiCache.get(markdown: emoji_token)
    _ ->
      :noop
  end
end)
```

Let me know your thoughts. For the most part, I think this explains how I was thinking of going about doing this.

Reply:
I don't think this is a bad idea. 

My issue is how we go about this with other modules. Take a module which has permissions for example. Undoubtedly we'll have some way of interacting with those, likely in a way in which supplying a struct wouldn't make sense. In that case, I don't think a similar approach would work.

What do you think?

Reply:
If I'm understanding correctly, I do not see this a problem. Discord permissions are usually bound to some kind of discord object (channel, role, overwrite, etc). I don't think there would be a scenario where we would have a cache that shares individual permissions. I may need more context.

Reply:
I wasn't thinking of needed a cache for the permissions. I meant to say that we'll probably want some helper methods around permissions, and I could see those taking in some extraneous parameters? I guess in that case you'd still want the discord object struct to perform the operation on, so it'd be okay.

I guess I can't think of anything off the top of my head as an example. In the case of permissions we might just want a different module that handles any logical operations on the subject? We can handle this when we get there.

Reply:
Yeah, I had the same idea with the permissions. Ideally, we'd create a module for the permissions that has functions for working with them. This would be similar to the `Nostrum.Struct.Snowflake` module.